module/zfs/spa_errlog.c

```diff
@@ -291,11 +287,23 @@ copyout_entry(const zbookmark_phys_t *zb, void *uaddr, uint64_t *count)
  */
 static int
 check_filesystem(spa_t *spa, uint64_t head_ds, zbookmark_err_phys_t *zep,
-    void *uaddr, uint64_t *count)
+    void *uaddr, uint64_t *count, int *zfs_checkfs_recursion)
 {
 	dsl_dataset_t *ds;
 	dsl_pool_t *dp = spa->spa_dsl_pool;
 
+	/*
+	 * The stack size of check_filesystem() is 208 bytes. For optimization
+	 * purposes multiply zfs_checkfs_recursion with 256 to get a rough
+	 * estimate of the stack depth. If this is within a page of the max
+	 * stack size return with an error. We return ECKSUM and not ENOMEM,
+	 * since the latter is also used when userspace needs to expand uaddr
+	 * in zpool_get_errlog().
+	 */
+	(*zfs_checkfs_recursion)++;
+	if ((*zfs_checkfs_recursion << 8) > (THREAD_SIZE - PAGE_SIZE))
+		return (ECKSUM);
```

Comment:
I have doubts about correctness of this 256byte assumption across different platforms, compilers and debug options.  On FreeBSD in several subsystems recursion depth is limited as:
`                /* Block direct execution if less then half of stack left. */
                size_t  st, su;
                GET_STACK_USAGE(st, su);
                if (su * 2 > st)
                        direct = 0;
`

Reply:
I will use another approach to iterate (not recursesively) over the clones using a list. That should avoid the stack overflow altogether.